### ISSUE TITLE
How to get requests working with jython 2.7.0?

### ISSUE THREAD

**Issue by 0Chencc** [jythontools/jython#83]
How to get requests working with jython 2.7.0?
![image](https://user-images.githubusercontent.com/19362246/28073529-8cbc2fd2-6688-11e7-98cf-bc03c080ce87.png)
![image](https://user-images.githubusercontent.com/19362246/28073585-c1da5996-6688-11e7-8818-02509322b4ad.png)
The requests folder already exists, but can not be called

**Comment by jeff5** [comment]
I suggest you install with pip. However, pip does not work properly in 2.7.0 on Windows (only Unices). Fortunately, Jython 2.7.1 is now out, where it does work.
https://mvnrepository.com/artifact/org.python/jython-installer/2.7.1

Make sure you use the pip that comes with Jython, not one you may have from CPython. This should result in requests being put in the Lib/site-packages folder, below where you installed Jython.

```
> java -jar D:\InstKits\jython.org\jython-installer-2.7.1.jar
Collecting setuptools
Collecting pip
Installing collected packages: setuptools, pip
Successfully installed pip-9.0.1 setuptools-28.8.0

> inst\bin\pip --version
pip 9.0.1 from C:\Users\Jeff\Documents\Jython\271-gh83\inst\Lib\site-packages (python 2.7)

> inst\bin\pip install requests
Collecting requests
  Downloading ...
Successfully installed certifi-2017.4.17 chardet-3.0.4 idna-2.5 requests-2.18.1 urllib3-1.21.1

> dir/w inst\Lib\site-packages
...
Directory of C:\Users\Jeff\Documents\Jython\271-gh83\inst\Lib\site-packages

[.]                           [..]                          [certifi]                     [certifi-2017.4.17.dist-info]
[chardet]                     [chardet-3.0.4.dist-info]     easy_install$py.class         easy_install.py
[idna]                        [idna-2.5.dist-info]          [pip]                         [pip-9.0.1.dist-info]
[pkg_resources]               README                        [requests]                    [requests-2.18.1.dist-info]
[setuptools]                  [setuptools-28.8.0.dist-info] [urllib3]                     [urllib3-1.21.1.dist-info]
...
> inst\bin\jython
Jython 2.7.1 (default:0df7adb1b397, Jun 30 2017, 19:02:43)
[Java HotSpot(TM) 64-Bit Server VM (Oracle Corporation)] on java1.8.0_131
Type "help", "copyright", "credits" or "license" for more information.
>>> import requests
>>> r = requests.get("http://jython.org")
>>> r.headers
{'Server': 'nginx', 'Connection': 'keep-alive', 'Last-Modified': 'Sun, 03 May 2015 07:46:38 GMT', 'Date': 'Sun, 16 Jul 2017 15:11:09 GMT', 'Accept-Ranges': 'bytes', 'ETag': '"5545d25e-4b0a"', 'Content-Length': '19210', 'Content-Type': 'text/html'}
```

**Comment by 0Chencc** [comment]
@jeff5 
C:\Users\0chen\Downloads
λ java -jar jython-install-2.7.1.jar
Error: Unable to access jarfile jython-install-2.7.1.jar

C:\Users\0chen\Downloads
λ java -jar jython-installer-2.7.1.jar
Exception:
Traceback (most recent call last):
  File "C:\Users\0chen\AppData\Local\Temp\tmpxfcf3l\pip-9.0.1-py2.py3-none-any.whl\pip\basecommand.py", line 215, in main
    status = self.run(options, args)
  File "C:\Users\0chen\AppData\Local\Temp\tmpxfcf3l\pip-9.0.1-py2.py3-none-any.whl\pip\commands\install.py", line 270, in run
    global_options = options.global_options or []
  File "C:\Users\0chen\AppData\Local\Temp\tmpxfcf3l\pip-9.0.1-py2.py3-none-any.whl\pip\basecommand.py", line 71, in _build_session
    retries=retries if retries is not None else options.retries,
  File "C:\Users\0chen\AppData\Local\Temp\tmpxfcf3l\pip-9.0.1-py2.py3-none-any.whl\pip\download.py", line 329, in __init__
    self.headers["User-Agent"] = user_agent()
  File "C:\Users\0chen\AppData\Local\Temp\tmpxfcf3l\pip-9.0.1-py2.py3-none-any.whl\pip\download.py", line 110, in user_agent
    if platform.system():
  File "D:\jython2.7.1\Lib\platform.py", line 1396, in system
    return uname()[0]
  File "D:\jython2.7.1\Lib\platform.py", line 1213, in uname
    machine = os.uname()[4]
IllegalArgumentException: java.lang.IllegalArgumentException: Cannot create PyString with non-byte value

**Comment by 0Chencc** [comment]
@jeff5 
How can i solve it?
Thanks!

**Comment by jeff5** [comment]
Darn. That's the result of characters > 255.

My guess is that os.uname() on its own will fail for you, and that this stems from a host name that contains Chinese characters. I can see exactly where that would bite in the code. We did a lot to make Jython 2.7.1 work with Chinese user names and file names. We didn't examine the  possibility that the _machine_ name could be Chinese. Would you be willing to change to an ascii-safe host name? I think that would get you past this point.

Am I right about the host name bing the cause? If so, we'll file a bug against that behaviour.

However, that even where/when Jython is fully competent, pip and installed packages may still contain an assumption that characters are bytes, especially in Python 2.

**Comment by 0Chencc** [comment]
@jeff5 
I use the Java call request, re module
Reported this error
```
Traceback (most recent call last):
   File "D: \ jython2.7.0 \ Lib \ encodings \ __ init__.py", line 31, in <module>
     Import codecs
   File "D: \ jython2.7.0 \ Lib \ codecs.py", line 268
     Return (b "", 0)
             ^
```
But in jython, but successfully run the py script.
How to solve.

**Comment by jeff5** [comment]
@0Chencc  I'm not quite sure what you are asking. However, I saw this error a lot with a Chinese localisation where the console encoding is code page 936. The cp936 codec is broken on 2.7.0. You could try setting codepage 1252 or 850 in your console.

But if we can get over the uname problem, I think 2.7.1 will work better for you. A lot of things are fixed in 2.7.1 that will be trouble for you in 2.7.0.

Did you understand my previous message about the machine name?

**Comment by jeff5** [comment]
Raised this as http://bugs.jython.org/issue2608 so we don't lose the host name issue, if genuine.

**Comment by jeff5** [comment]
I raised the host name issue as http://bugs.jython.org/issue2608 so we don't lose it. I can reproduce it by setting the host name to 先知_MICAH, so a work-around is to change the computer name to something Western-European. (Sorry, I realise it shouldn't be necessary.)

**Comment by 0Chencc** [comment]
I installed 2.7.1 without PIP

**Comment by jeff5** [comment]
Good work-around (if you don't need pip). But don't you need it to install requests?

It was fairly painless for me to change the host name, so you could do it temporarily, if you have the privilege to do so. Or maybe install on another machine then move the installation? There are no registry changes AFAIK - just the files in the installation directory.

This bug is fixed in the codebase, but there's no schedule yet for 2.7.2.

**Comment by devendor** [comment]
I have a series of patches I'm working on that should fix this.  Not quite ready for prime time yet.  Taken together, they get the latest pip working which itself relies on urllib3.

**Comment by mjcumming** [comment]
@devendor I am  very interested in your solution.  

Mike

**Comment by jeff5** [comment]
@devendor, @mjcumming : the bug reported here is closed, fixed. If it is necessary to fix something else with pip, that's a distinct issue, probably http://bugs.jython.org/issue2660 .

**Comment by shelja01** [comment]
How was this issue resolved? I am interested in the solution, facing the same issue.

**Comment by jeff5** [comment]
There are several issues in this thread, first pip in 2.7.0 on Windows, then intolerance in 2.7.1 of a non-ascii hostname, which should be fixed in 2.7.2 (now in beta). Work-around may have been don't install pip, or change the host name. The urllib3 problem may be yet another.

So when you say "the same issue ..."?
